### PR TITLE
Rpc/declare transaction v1

### PR DESCRIPTION
**Pull request by marioiordanov**
## Usage related changes

RPC implementation for Declare V1

## Checklist:

- [x] Applied formatting - `./scripts/format.sh`
- [x] No linter errors - `./scripts/clippy_check.sh`
- [x] Performed code self-review
- [x] Rebased to the last commit of the target branch (or merged it into my branch)
- [ ] Documented the changes
- [x] Linked the issues which this PR resolves
- [ ] Updated the tests
- [x] All tests are passing - `cargo test`

### CODE REVIEW

File: crates/starknet-server/src/api/json_rpc/write_endpoints.rs [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+        assert_eq!(
+            result.class_hash.0.to_prefixed_hex_str(),
+            "0x399998c787e0a063c3ac1d2abac084dcbe09954e3b156d53a8c43a02aa27d35"
+        );
+
+        println!("{}", result.transaction_hash.0.to_prefixed_hex_str());
+    }
+
+    fn setup() -> JsonRpcHandler {
+        let config = StarknetConfig {
```
Comment (FabijanC) [inline crates/starknet-server/src/api/json_rpc/write_endpoints.rs:119]: Should we use the new `starknet_config_for_test` here?

File: crates/starknet-server/src/api/json_rpc/write_endpoints.rs [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+        let starknet = Starknet::new(&config).unwrap();
+        let api = Api::new(starknet);
+        JsonRpcHandler { api }
+    }
+
+    #[test]
+    fn parsed_base64_gzipped_json_contract_class_correctly_computed_hash() {
+        let json_string = std::fs::read_to_string(concat!(
+            env!("CARGO_MANIFEST_DIR"),
+            "/test_data/rpc/declare_v1.json"
```
Comment (FabijanC) [inline crates/starknet-server/src/api/json_rpc/write_endpoints.rs:133]: And what are we declaring? Should it be documented somewhere? Right now it's a hardcoded artifact and it is not clear what is the source.

**Review by FabijanC — CHANGES_REQUESTED** [review]

File: crates/starknet/src/services/add_declare_transaction.rs [hunk 3]
```diff
@@ -292,6 +301,24 @@ mod tests {
...
+
+        let err = starknet.add_declare_transaction_v1(declare_txn).err().unwrap();
+
+        if let starknet_types::error::Error::TransactionError(
+            TransactionError::ClassAlreadyDeclared(_),
+        ) = err
+        {
+        } else {
+            panic!("Wrong error type");
+        }
```
Comment (FabijanC) [inline crates/starknet/src/services/add_declare_transaction.rs:319]: I spent too much time deciphering this, wouldn't the plain old `match` be clearer? 
```suggestion
        match err {
            starknet_types::error::Error::TransactionError(
                TransactionError::ClassAlreadyDeclared(_),
            ) => (),
            _ => panic!("Wrong error type"),
        }
```

File: crates/starknet-server/src/api/json_rpc/write_endpoints.rs [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+    type Error = ApiError;
+    fn try_from(value: BroadcastedDeclareTransactionV1) -> RpcResult<Self> {
+        Ok(DeclareTransactionV1 {
+            sender_address: value.sender_address.0,
+            max_fee: value.common.max_fee.0,
+            signature: value.common.signature.iter().map(|x| x.0).collect(),
+            nonce: value.common.nonce.0,
+            contract_class: ContractClass::try_from(value.contract_class)?,
+            class_hash: None,
+            transaction_hash: None,
```
[the remark this answers is not in the document]
Comment (marioiordanov) [reply crates/starknet-server/src/api/json_rpc/write_endpoints.rs]: Resolved in [7b19c15](https://github.com/0xSpaceShard/starknet-devnet-rs/pull/39/commits/7b19c15c3046129cf57226283dcaac5896e442df)

File: crates/starknet-server/src/api/json_rpc/write_endpoints.rs (hunk 2 above)
Comment (marioiordanov) [reply crates/starknet-server/src/api/json_rpc/write_endpoints.rs:134]: Left a comment in [7c72a10](https://github.com/0xSpaceShard/starknet-devnet-rs/pull/39/commits/7c72a107d1376d8af48e814da62c52f4af76fd26)

File: crates/starknet/src/services/add_declare_transaction.rs (hunk 3 above)
Comment (marioiordanov) [reply crates/starknet/src/services/add_declare_transaction.rs]: Removed the test in [b1dfbba](https://github.com/0xSpaceShard/starknet-devnet-rs/pull/39/commits/b1dfbba19216a4e961fa2719ca41f68a833ec883)

File: crates/starknet/src/services/add_declare_transaction.rs [hunk 5]
```diff
@@ -93,6 +94,14 @@ impl Starknet {
         let transaction_hash = declare_transaction.generate_hash()?;
         declare_transaction.transaction_hash = Some(transaction_hash);
 
+        if self.state.is_contract_declared(&class_hash)? {
```
[the remark this answers is not in the document]
Comment (marioiordanov) [reply crates/starknet/src/services/add_declare_transaction.rs]: Removed in [b1dfbba](https://github.com/0xSpaceShard/starknet-devnet-rs/pull/39/commits/b1dfbba19216a4e961fa2719ca41f68a833ec883)

File: crates/starknet-server/src/api/json_rpc/write_endpoints.rs (hunk 1 above)
Comment (FabijanC) [reply crates/starknet-server/src/api/json_rpc/write_endpoints.rs:117]: Ok, I didn't realize that, let's keep it as it is.

File: crates/starknet/src/services/add_declare_transaction.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (FabijanC) [reply crates/starknet/src/services/add_declare_transaction.rs]: Did you check this or did you just believe my word?

[the remark this answers is not in the document]
Comment (marioiordanov) [reply crates/starknet/src/services/add_declare_transaction.rs]: Believe your word

[the remark this answers is not in the document]
Comment (FabijanC) [reply crates/starknet/src/services/add_declare_transaction.rs]: Then let's not merge until I check this.

[the remark this answers is not in the document]
Comment (FabijanC) [reply crates/starknet/src/services/add_declare_transaction.rs]: Just checked on the integration network, redeclaration was successful - no errors reported.

**Review by FabijanC — APPROVED** [review]

**Pull request merged**